src/main/scala/viper/gobra/frontend/Desugar.scala

```diff
@@ -1467,20 +1577,99 @@ object Desugar {
                 convertedArgs = convertArgs(args)
                 mproxy = getMethodProxy(base, recv, convertedArgs)
-              } yield Right(in.PureMethodCall(recv, mproxy, convertedArgs, resT)(src))
+                spec = p.maybeSpec.map(closureSpecD(ctx, info))
+              } yield Right(pureMethodCall(recv, mproxy, convertedArgs, spec, resT))
             } else {
               for {
                 (recv, args) <- dRecvWithArgs
                 convertedArgs = convertArgs(args)
                 mproxy = getMethodProxy(base, recv, convertedArgs)
-              } yield Left((targets, in.MethodCall(targets, recv, mproxy, convertedArgs)(src)))
+                spec = p.maybeSpec.map(closureSpecD(ctx, info))
+              } yield Left((targets, methodCall(targets, recv, mproxy, convertedArgs, resT, spec)))
             }
           }
           case sym => Violation.violation(s"expected symbol with arguments and result or a built-in entity, but got $sym")
         }
       }
     }
 
+    /** Returns either a tuple with targets and call statement or, if the call is pure, the call expression directly. */
+    def closureCallDAux(ctx: FunctionContext, info: TypeInfo)(expr: PInvoke)(src: Meta): Writer[Either[(Vector[in.LocalVar], in.Stmt), in.Expr]] = {
+      val (func, isPure) = info.resolve(expr.spec.get.func) match {
+        case Some(ap.Function(_, f)) => (f, f.isPure)
+        case Some(ap.Closure(_, c)) => (c, c.isPure)
+        case _ => violation("expected function or function literal")
+      }
+
+      val spec = closureSpecD(ctx, info)(expr.spec.get)
+      val params = func.args.zipWithIndex.collect { case (p, idx) if !spec.params.contains(idx + 1) => func.context.typ(p) }
+
+      if (isPure) for {
+        dArgs <- functionCallArgsD(expr.args, params)(ctx, info, src)
+        args = arguments(dArgs zip params)
+        closure <- exprD(ctx, info)(expr.base.asInstanceOf[PExpression])
+        resT = typeD(func.context.typ(func.result), Addressability.callResult)(src)
+      } yield Right(in.PureClosureCall(closure, arguments(args zip params), spec, resT)(src))
+      else for {
+        dArgs <- functionCallArgsD(expr.args, params)(ctx, info, src)
+        args = arguments(dArgs zip params)
+        closure <- exprD(ctx, info)(expr.base.asInstanceOf[PExpression])
+        targets = func.result.outs map (o => freshExclusiveVar(typeD(func.context.symbType(o.typ), Addressability.exclusiveVariable)(src), expr, info)(src))
+      } yield Left((targets, in.ClosureCall(targets, closure, arguments(args zip params), spec)(src)))
+    }
+
+
+    /** Desugars the arguments to a function call or a call with spec.
+      * @param args The expressions passed as arguments
+      * @param params The types of the arguments of the callee (or the spec) */
+    private def functionCallArgsD(args: Vector[PExpression], params: Vector[Type])(ctx: FunctionContext, info: TypeInfo, src: Meta): Writer[Vector[in.Expr]] = {
```

Comment:
Here you moved code from the `funcationCallAux` method, right? This refactoring seems like a good ideal.

Reply:
Yes